diff --git a/node_modules/sass-loader/dist/utils.js b/node_modules/sass-loader/dist/utils.js
index cd50f34..3bee594 100644
--- a/node_modules/sass-loader/dist/utils.js
+++ b/node_modules/sass-loader/dist/utils.js
@@ -93,6 +93,9 @@ function getSassImplementation(loaderContext, implementation) {
     } // eslint-disable-next-line consistent-return
 
 
+    return resolvedImplementation;
+  } else if (implementationName === "sass-embedded") {
+    // eslint-disable-next-line consistent-return
     return resolvedImplementation;
   }
 
